src/feat/wave-reader.cc

```diff
@@ -221,55 +221,88 @@ void WaveData::Read(std::istream &is) {
   uint32 data_chunk_size = ReadUint32(is, swap);
   riff_chunk_read += 4;
 
+  // check if data is read in stream mode
+  bool is_stream_mode = false;
+  if ((riff_chunk_size == 0 && data_chunk_size == 0)
+        || (riff_chunk_size == std::numeric_limits<uint32>::max() 
+            && data_chunk_size == std::numeric_limits<uint32>::max())) {
+    // when data are read in stream model, we don't know chunk size in advance
+    // data_chunk_size and riff_chunk_size (Uint32) could be 
+    // either ox00000000 or oxFFFFFFFF
+    is_stream_mode = true;
+  }
+
   if (std::abs(static_cast<int64>(riff_chunk_read) +
                static_cast<int64>(data_chunk_size) -
                static_cast<int64>(riff_chunk_size)) > 1) {
     // we allow the size to be off by one, because there is a weirdness in the
     // format of RIFF files that means that the input may sometimes be padded
     // with 1 unused byte to make the total size even.
-    KALDI_ERR << "Expected " << riff_chunk_size << " bytes in RIFF chunk, but "
-              << "after first data block there will be " << riff_chunk_read
-              << " + " << data_chunk_size << " bytes "
-              << "(we do not support reading multiple data chunks).";
+    if (is_stream_mode) 
+        KALDI_WARN << "Read in RIFF chunk size: " << riff_chunk_size
+                   << ", Read in data chunk size: " << data_chunk_size
+                   << ". Maybe data is read in stream mode "
+                   << "so that we don't know exact size in advance.";
+    else
+        KALDI_ERR << "Expected " << riff_chunk_size << " bytes in RIFF chunk, but "
+                  << "after first data block there will be " << riff_chunk_read
+                  << " + " << data_chunk_size << " bytes "
+                  << "(we do not support reading multiple data chunks).";
   }
 
-  std::vector<char*> data_pointer_vec;
-  std::vector<int> data_size_vec;
+  std::vector<char> chunk_data_vec;
   uint32 num_bytes_read = 0;
-  for (int32 remain_chunk_size = data_chunk_size; remain_chunk_size > 0;
-       remain_chunk_size -= kBlockSize) {
-    int32 this_block_size = remain_chunk_size;
-    if (kBlockSize < remain_chunk_size)
-      this_block_size = kBlockSize;
-    char *block_data_vec = new char[this_block_size];
-    is.read(block_data_vec, this_block_size);
-    num_bytes_read += is.gcount();
-    data_size_vec.push_back(is.gcount());
-    data_pointer_vec.push_back(block_data_vec);
-    if (num_bytes_read < this_block_size)
-      break;
-  }
-
-  std::vector<char> chunk_data_vec(num_bytes_read);
-  uint32 data_address = 0;
-  for (int i = 0; i < data_pointer_vec.size(); i++) {
-    memcpy(&(chunk_data_vec[data_address]), data_pointer_vec[i],
-           data_size_vec[i]);
-    delete[] data_pointer_vec[i];
-    data_address += data_size_vec[i];
+  if (!is_stream_mode) {
```

Comment:
Can you maybe rewrite this branch as well? I understand it is old code, but it is extremely ugly. No need to allocate a kerzillion of small buffers, read into them, then merge into a large one. Just read all into its final location in one single read! You should actually get 3 lines of code instead of 40 some. Something to the effect of

```C++
chunk_data_vec.resize(data_chunk_size);
is.read(&chunk_data_vec.first(), data_chunk_size);
uint32 num_bytes_read = is.gcount();
```

This is actually all it should take to read the chunk!